pkg/util/volumeclaimtemplate/volume_templates_hash.go

```diff
@@ -0,0 +1,136 @@
+	// deprecated: can not cover any scenario, some pvc type will fill some fields in pvc field however not found in vcTemplate
+	GetCurrentHash(volumes []v1.Volume, getter PVCGetter) (uint64, error)
+}
+
+type PVCGetter = func(string) *v1.PersistentVolumeClaim
+
+type volumeClaimTemplatesHasher struct{}
+
+func NewVolumeClaimTemplatesHasher() VolumeClaimTemplatesHasher {
+	return &volumeClaimTemplatesHasher{}
+}
+
+func (h *volumeClaimTemplatesHasher) GetExpectHash(templates []v1.PersistentVolumeClaim) uint64 {
+	var pvcs []v1.PersistentVolumeClaim
+	for i := range templates {
+
+		pvcs = append(pvcs, v1.PersistentVolumeClaim{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: templates[i].Name,
+			},
+			Spec: templates[i].Spec,
+		})
+	}
+
+	sort.SliceStable(pvcs, func(i, j int) bool {
+		return pvcs[i].Name < pvcs[j].Name
+	})
+	return hashVolumeClaimTemplate(pvcs)
+}
+
+func (h *volumeClaimTemplatesHasher) GetCurrentHash(volumes []v1.Volume, getter PVCGetter) (uint64, error) {
+	var pvcs []v1.PersistentVolumeClaim
+	for i := range volumes {
+		if volumes[i].PersistentVolumeClaim == nil {
+			continue
+		}
+		pvcName := volumes[i].PersistentVolumeClaim.ClaimName
+		pvc := getter(volumes[i].PersistentVolumeClaim.ClaimName)
+		if pvc == nil {
+			// error here
+			return 0, fmt.Errorf("pvc %v not exist", pvcName)
+		}
+
+		pvcs = append(pvcs, v1.PersistentVolumeClaim{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: volumes[i].Name,
+			},
+			Spec: pvc.Spec,
+		})
+	}
+
+	sort.SliceStable(pvcs, func(i, j int) bool {
+		return pvcs[i].Name < pvcs[j].Name
+	})
+	return hashVolumeClaimTemplate(pvcs), nil
+}
+
+func hashVolumeClaimTemplate(templates []v1.PersistentVolumeClaim) uint64 {
+	hash := fnv.New32a()
+	envsJSON, _ := json.Marshal(templates)
+	hashutil.DeepHashObject(hash, envsJSON)
+	return uint64(hash.Sum32())
+}
+
+func GetVCTemplatesHash(revision *apps.ControllerRevision) (string, bool) {
+	if len(revision.Annotations) > 0 {
+		val, exist := revision.Annotations[VolumeClaimTemplateHashAnnotation]
+		return val, exist
+	}
+	return "", false
+}
+
+func PatchVCTemplateHash(revision *apps.ControllerRevision, vcTemplates []v1.PersistentVolumeClaim) {
+	if len(revision.Annotations) == 0 {
+		revision.Annotations = make(map[string]string)
+	}
+	if len(vcTemplates) != 0 {
+		// get hash of vct
+		vcTemplateHash := NewVolumeClaimTemplatesHasher().GetExpectHash(vcTemplates)
+		revision.Annotations[VolumeClaimTemplateHashAnnotation] = strconv.FormatUint(vcTemplateHash, 10)
+	} else {
+		revision.Annotations[VolumeClaimTemplateHashAnnotation] = ""
+	}
+}
+
+func CanVCTemplateInplaceUpdate(oldRevision, newRevision *apps.ControllerRevision) bool {
+	newVCTemplatesHash, newExist := GetVCTemplatesHash(newRevision)
+	oldVCTemplatesHash, oldExist := GetVCTemplatesHash(oldRevision)
+	// oldExist = false, pass this case to avoid unnecessary recreate when first open this feature
+	// newExist = false, actually this case not exist
```

Comment:
If this is not the case, isn't it unnecessary to judge newExist?

Reply:
re-add this judgement again.

One corner case may exist:
1. cloneset A with image 1 vct size 1Gi
2. kriuse update to new version, next revision will patch hash
3. 1st update cloneset A with image2 vct size 1Gi -> in-place update and revision hash is not nil
4. 1st update over
5. kruise rollback, next revision will not patch hash
6. 2rd update cloneset A with image3 vct size 1Gi, partition 50% -> in-place update and revision hash is nil
7. kruise re-update
8. 2rd update, with partition 0 

In step8 of this case, new revision hash does not exist and old revision exists.
It might be necessary to include a case where the new revision is non-existent as well.